### DISCUSSION TITLE
iPhones running iOS 17 and above and private 5G networks

### DISCUSSION THREAD

**Discussion by herlesupreeth** [herlesupreeth/docker_open5gs#248] [Announcements]
This is just an information post regarding testing private 5G SA with iPhones.

Here is the webpage outlining the requirements from Apple - https://support.apple.com/en-ie/guide/deployment/depac6747317/web

Following steps needs to be done to address requirement for iPhones to attach to private 5G SA:

1. SUPI concealment when exchanging over 5G network. In order to achieve that certain 5G files in SIM card needs to be programmed. A pysim script is provided at https://github.com/herlesupreeth/docker_open5gs/blob/master/sim/iphone-private-5g.script to program your Sysmocom SIM card with relevant files for the **UDM configured in this repository** (Dont forget to change the ADM key as per your SIM card in that script before executing). You have to execute as follows:

```
./pySim-shell.py -p 0 --script scripts/deactivate-5g.script
./pySim-shell.py -p 0 --script scripts/iphone-private-5g.script
```

2. Enable **NAS encryption**: This can be done with following changes in `amf.yaml` in `amf` folder of this repository

```
    security:
        integrity_order : [ NIA2, NIA1, NIA0 ]
        ciphering_order : [ NEA2, NEA1, NEA0 ]
```

3. Enable **User plane encryption** at gNB: This can be done by passing the following command line args to srsRAN_Project gNB:

```
cu_cp security --nea_pref_list=nea2,nea1,nea3,nea0
```

4. Also, you would have to use the PLMN 999 xx assigned for private 5G network

**After inserting the SIM card goto Settings -> Cellular -> Voice & Data and enable "5G Standalone" option**

I have tested iPhone 13 Pro and verified the attach with srsRAN_Project gNB.

NOTE: I recently discovered that when operating using a private PLMN (e.g. 999 XX), Apple restricts attaching of iPhone to network operating only on NR bands which are allocated for private networks in respective countries. For example, if you are in USA you can use Band 41 when using private network PLMN.

**Comment by herlesupreeth** [comment]
NOTE: The above requirement of NAS encryption and User plane encryption is not necessary when testing with 001 01 test PLMN. The only requirement is the availability of IMS and SUPI concealment. i.e. **skip NAS encryption and User plane encryption mentioned above**

**iPhone blacklists SIM's ICCID from connecting to 5G SA if the PLMN is not one of the PLMN assigned for private mobile network**

Luckily, the above blacklisting can be circumvented as follows:
1. First program the SIM with PLMN corresponding to a private mobile network PLMN e.g. 999 70 
2. Start the 5GC with the SIM programmed PLMN
3. Start the gNB with the SIM programmed PLMN
4. Put the SIM in iPhone and attempt to attach to gNB.
5. Enable **5G Standalone** option in **Voice & Data** settings menu. **Wait for UE to attach to network**
6. Then, bring down the network and put phone in airplane mode and pull out the SIM
7. Re-program the SIM to any desired PLMN e.g. 001 01
8. Now, iPhone allows attaching to 5G SA network with non-private mobile network PLMN

When testing with test PLNM (001 01), iPhone 13 pro was able to connect to `default` and `ims` APN and register with IMS successfully. 

**You would need to use the following deployment file https://github.com/herlesupreeth/docker_open5gs/blob/master/sa-vonr-deploy.yaml to have a successful connection**

**Reply by platini2** [reply]
HI quick question, where do i derive the public key of SUCI in the simcard from my deployment?

**Reply by herlesupreeth** [reply]
If you are using this repo, then the key is already in this file https://github.com/herlesupreeth/docker_open5gs/blob/master/sim/iphone-private-5g.script

If not, then you need to replace the contents of "hnet_pubkey" with the public key you find at the location provided by the below configuration in udm.yaml

https://github.com/herlesupreeth/docker_open5gs/blob/9c034db839a76c31821c752f93b92793d2659bce/udm/udm.yaml#L15

In order to extract the public key you need to execute below command

`openssl pkey -in curve25519-1.key -text`

**Reply by platini2** [reply]
Thanks! i just found after i asked the question. there is a good guide from @s5uishida to generate them at
https://github.com/s5uishida/note_5g_suci_profile_ab

**Reply by lmesserStep** [reply]
Hey @herlesupreeth what is stopping me from using 3150-10 (Non-Public Network) on an iPhone 14? I am generating an eSIM for this. 

I'm having trouble understanding why I would need IMS, Apple's official documentation seems to say a data-only network is fine.

**Reply by herlesupreeth** [reply]
@lmesserStep If you are using private network PLMN you dont need IMS.

**Reply by andymorris328** [reply]
Here is recent data point:
 
We see very stable iPhone data connections on Private Networks PLM 999 99 with iPhone 16 and 16 Pro following the advice in this project:

Steps to follow:
- SIM in main SIM tray: Operating as primary SIM, or operating as secondary SIM if public MNO eSIM is already installed
- SIM programmed with PLMN 99 xx (otherwise iPhone does not scan for private network frequencies and will not detect the cell - we used 999 99 for testing). 
- SIM/Core configured for SUPI concealment
- NAS encryption NEA2
- Enable 5G SA switch in mobile settings
- Configured DNNs:
   Data: internet  
   **VoNR not required**: ims (iPhone never attempted to register for ims apn even if provided in settings, check details in published Apple tech note about restriction of voice calling over private networks)

**Reply by KnechtNoobrecht** [reply]
@andymorris328 what if 5G SA switch is deactivated with comment "SIM does not support 5G SA"? using sysmocom SIMs

**Comment by gabri94** [comment]
Hi All, we've been trying for a while to use these instructions to generate a sim card that can connect to our private 5G network on PLMN 99901, however our Iphone 14 pro is never able to detect the network.

We are able to detect and register the network (with PLMN 00101)  if we use the following SIM card configurations:
- opencell sim card with 00101 (no supi/suci)
- sysmocom simcard with 00101 (no supi/suci)
- sysmocom simcard with 00101 (with supi/suci, configured as above)

We also tested using a Sierra 5G Modem and we are able to detect/register to our 99901 network using a simcard programmed for 99901 w/o suci.

Any idea?

**Reply by platini2** [reply]
Pretty much on the same boat, have not been able to detect network on an iphone 13 pro max with 99901 either with SUPI or without. A samsung UE registers with no problem. Bands 78 and 66.

**Reply by herlesupreeth** [reply]
I recently discovered that when operating using a private PLMN (e.g. 999 XX), Apple restricts attaching of iPhone to network operating only on NR bands which are allocated for private networks in respective countries. For example, if you are in USA you can use Band 41 when using private network PLMN.

**Reply by gabri94** [reply]
Thanks @herlesupreeth for this insight. Is there a list of bands allowed per country? My understanding is that n48 (cbrs) was allocate for private 5G in the US

 We're in the US and have been testing with n78, we'll try with both n41 and n48.

**Reply by herlesupreeth** [reply]
> Is there a list of bands allowed per country?

I couldn't find a common source where it lists bands allocated for private network.

**Reply by platini2** [reply]
Great information @herlesupreeth, will try to get it up with band 41.

**Reply by gabri94** [reply]
We were able to register our EU iPhone 14 pro (in the US)  using n77 and n41. 
With n78 and n48 we were able to detect the network, but iPhone was not registering

**Reply by platini2** [reply]
> We were able to register our EU iPhone 14 pro (in the US) using n77 and n41.
With n78 and n48 we were able to detect the network, but iPhone was not registering

What gnb sw are you using? OAI or SRSRAN_5g?

**Reply by gabri94** [reply]
We tested both OAI+USRP and Nvidia ARC+Foxconn RU.

**Reply by platini2** [reply]
Thanks finally i was able to get the Iphone 14 Pro MAX in n78 SA with OAI+USRP running on a Raspberry PI 5 with 51PRBs

**Reply by shivakumardistproc** [reply]
> I recently discovered that when operating using a private PLMN (e.g. 999 XX), Apple restricts attaching of iPhone to network operating only on NR bands which are allocated for private networks in respective countries. For example, if you are in USA you can use Band 41 when using private network PLMN.

1. Do Pixel phones have similar restrictions? Specifically:
       a. Do private PLMNs work only on NR bands allocated for private networks in their respective countries?
       b. More generally, does a Pixel phone restrict NR bands for a given PLMN?
2. Does anyone know how Pixel phones apply carrier configurations?

**Reply by gabri94** [reply]
@shivakumardistproc I believe so, but it's almost impossible to find details about it. There are some Pixel-specific carrier configurations that I found about in a telegram channel

**Reply by L-CHINA** [reply]
> Thanks finally i was able to get the Iphone 14 Pro MAX in n78 SA with OAI+USRP running on a Raspberry PI 5 with 51PRBs

can you help me？my Iphone 15 pro is never able to detect the network.

**Comment by gabri94** [comment]
Hi All, was anyone able to use Android phones with non-00101 PLMN? 
We've been testing some Pixel phones, but they don't seem to have 5G SA enabled with 999-xx PLMNS

**Reply by platini2** [reply]
I have a Samsung Galaxy A15 5G registered in SA with PLMN 999-01, working even with and without SUPI concealment. Band 78.

**Reply by gabri94** [reply]
I confirm that on a Samsung S23 the phone can connect right away to 99901, however on both Pixel 8pro, Pixel 6 and Oneplus Nord the PLMN seems to be somehow locked on the modem.

**Reply by ffxxcad** [reply]
I have seen the same issue with Pixel 7a

**Reply by gabri94** [reply]
Same with Pixel 9 pro.

**Reply by herlesupreeth** [reply]
@gabri94 have you tried running CoIMS? I had added a configuration which tells the network as having 5G SA capability

**Reply by gabri94** [reply]
I installed it, but I don't have carrier privileges apparently. Will try to follow the guide in the next days.

Can we also use it with our Private Networ eSIM ? (we have all the key material)

**Reply by herlesupreeth** [reply]
> Can we also use it with our Private Networ eSIM ? (we have all the key material)

It should be possible if eSIM support OTA based RFM (to install the applet and push SHA1 key). I haven't tested it so cant say for sure

**Reply by gabri94** [reply]
I was able to get CoIMS working by loading the Private eSIM (999-02) in a physical eUICC from sysmocom. 
What's the config line I should look for in CoIMS? There's so many of them

**Reply by herlesupreeth** [reply]
> What's the config line I should look for in CoIMS? There's so many of them

you just need to run the CoIMS app once and it will set all the required configuration to enable VoLTE. Basically, the following setting must be displayed as true (its its not try pressing REFRESH and then it should show)

carrier_volte_provisioned_bool = true;

**Reply by gabri94** [reply]
Yeah, I noticed that a lot of options have appeared in the phone now. Unfortunately, when I enter a [e]SIM with PLMN 999**, the phone is still locked in LTE mode. It won't get in 5GNR even by forcing it.

**Reply by herlesupreeth** [reply]
Have you tried changing NR bands based on your region as mentioned in this comment ? - https://github.com/herlesupreeth/docker_open5gs/discussions/248#discussioncomment-11217257

**Reply by gabri94** [reply]
I think we're stuck a step behind that. We can see with Network Signal Guru that the modem is stuck in LTE mode.
However if we use a SIM card with PLMN 00101, the modem correctly goes in 5GNR mode. 

I think somehow the Shannon modem is restricting the 5GNR mode to certain PLMNs.

**Reply by andymorris328** [reply]
It is possible to successfully test VoNR on Private Network bands with PLMN 999 99 using COTS Samsung S24 One UI 7 with no CoIMS or UE diagnostic software.

Steps:
1. **Use SIM slot 2** (if you use SIM slot 1, it appears that the handset does not scan for private network bands)
2. _*_#_*_#4636#_*#_* and select NR only (this speeds up the network selection)
3. Configured DNNs: 
    Data: internet, type=default, ipv4. 
    VoNR: ims, type=ims, ipv4
4. Enable VoLTE calling in the settings if it not enabled 
* SIM is configured for SUPI concealment and NAS encryption is enabled. Both are needed for iPhone but shouldn't be required for Samsung. IMPU/IMPI is also correctly configured.

Trying above steps with S21 One UI 7 (latest software at time) does not result in the same success. S21 only attaches to internet dnn and fails to attempt to connect to the ims dnn, as a result only data sessions are possible. (Caveat being that this device is running diagnostic software and may behave slightly differently from COTS).

Pixel 9 Pro XL tested with same SIM in the only SIM slot and it can't detect the cell. (Not tried with eSIM yet)

**Reply by gabri94** [reply]
I will give it a try with the Sim Slot trick, Here I were also able to have certain Samsung  phones S23/24[Ultra] to connect using a physical SIM. SUCI doesn't seem to required.

As per the Pixel, I found out that the culprit is the carrierconfig db. (`/vendor/firmware/carrierconfig/cfg.db`). This file maps which PLMN are allowed to connect using 5GSA and each PLMN must be enabled manually. With a Magisk patch (need a rooted phone) I was able to override the DB and enable PLMN 99901 on Pixel 6.
Unfortunately there doesn't seem to be a way without root.

**Reply by ankithmr** [reply]
@gabri94 , what's the make of your s24 Ultra. I have a US made S24 and it doesn't work with 00101 or 99999. Any tricks ?   On the other hand s24 just works fine with both PLMNs

**Comment by luckybird77** [comment]
@herlesupreeth I followed your instructions on iPhone 14 and it works wonderfully for 5G NR Band 41.  I used PLMN 999-70.  However, when trying to use 5G NR n77 Band, the phone does not even make an attempt to attach.  I also tried with 00101 but there is no success.  Do you have any insight to this ?   I also tried dummy PLMN 302 995 (open for use for testing in canada), but nothing seems to work. I know the phone supports n77 band and my Android phone can attach to n77 with PLMN 00101

**Reply by herlesupreeth** [reply]
When you change band you would need to manually search for networks for the very first time. Have you done this?

if phone does not detect the network even after above step then I would check whether the bandwidth you are using on band 77 is supported or not. This you can verify by attaching to band 41 like before and checking UE capabilities information

**Comment by luckybird77** [comment]
Hi Herle,
Thank you for your reply.

Yes, I have manually search but it is not finding it.
How do I check the UE capabilities in iphone ? In band 41 it was able to
attach with 20 MHz

Hendry

**Reply by herlesupreeth** [reply]
When you attach the iPhone to band 41, check the pcap on the gNB (if you are using SRS RAN, you should be able to see it in f1ap pcap). If you post here the pcap I can take a look. Also, please mention which bandwidth are you testing with band 77

**Reply by luckybird77** [reply]
Thank you for your help  .. I was able to capture the UE capabilities

grep -i -C 5 "bandNR" open5gs_ngap_dump.txt 
```

pdsch-RE-MappingFR1-PerSymbol: n10 (0)
pdsch-RE-MappingFR1-PerSlot: n32 (1)
rf-Parameters
supportedBandListNR: 10 items
Item 0
BandNR
bandNR: 78
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
scs-30kHz: 8000 [bit length 16, 1000 0000  0000 0000 decimal value 32768]
channelBWs-UL-v1590: fr1 (0)
fr1
scs-30kHz: 8000 [bit length 16, 1000 0000  0000 0000 decimal value 32768]
Item 1
BandNR
bandNR: 41
modifiedMPR-Behaviour: c0 [bit length 8, 1100 0000 decimal value 192]
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
--
scs-30kHz: 8000 [bit length 16, 1000 0000  0000 0000 decimal value 32768]
channelBWs-UL-v1590: fr1 (0)
fr1
scs-30kHz: 8000 [bit length 16, 1000 0000  0000 0000 decimal value 32768]
Item 2
BandNR
bandNR: 71
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
fr1
scs-15kHz: f000 [bit length 10, 6 LSB pad bits, 1111 0000  00.. .... decimal value 960]
scs-30kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
scs-60kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
Item 3
BandNR
bandNR: 66
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
scs-15kHz: f400 [bit length 10, 6 LSB pad bits, 1111 0100  00.. .... decimal value 976]
scs-30kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
scs-60kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
asymmetricBandwidthCombinationSet: 80 [bit length 1, 7 LSB pad bits, 1... .... decimal value 1]
Item 4
BandNR
bandNR: 25
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
fr1
scs-15kHz: f000 [bit length 10, 6 LSB pad bits, 1111 0000  00.. .... decimal value 960]
scs-30kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
scs-60kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
Item 5
BandNR
bandNR: 7
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
fr1
scs-15kHz: f000 [bit length 10, 6 LSB pad bits, 1111 0000  00.. .... decimal value 960]
scs-30kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
scs-60kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
Item 6
BandNR
bandNR: 5
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
fr1
scs-15kHz: f000 [bit length 10, 6 LSB pad bits, 1111 0000  00.. .... decimal value 960]
scs-30kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
scs-60kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
Item 7
BandNR
bandNR: 2
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
fr1
scs-15kHz: f000 [bit length 10, 6 LSB pad bits, 1111 0000  00.. .... decimal value 960]
scs-30kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
scs-60kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
Item 8
BandNR
bandNR: 29
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
fr1
scs-15kHz: c000 [bit length 10, 6 LSB pad bits, 1100 0000  00.. .... decimal value 768]
scs-30kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
scs-60kHz: 0000 [bit length 10, 6 LSB pad bits, 0000 0000  00.. .... decimal value 0]
Item 9
BandNR
bandNR: 12
mimo-ParametersPerBand
tci-StatePDSCH
maxNumberConfiguredTCIstatesPerCC: n16 (2)
maxNumberActiveTCI-PerBWP: n1 (0)
pusch-TransCoherence: nonCoherent (0)
--
BandCombination
bandList: 1 item
Item 0
BandParameters: nr (1)
nr
bandNR: 41
ca-BandwidthClassDL-NR: c (2)
ca-BandwidthClassUL-NR: a (0)
featureSetCombination: 0
supportedBandwidthCombinationSet: c0 [bit length 2, 6 LSB pad bits, 11.. .... decimal value 3]
Item 1
BandCombination
bandList: 2 items
Item 0
BandParameters: nr (1)
nr
bandNR: 41
ca-BandwidthClassDL-NR: a (0)
ca-BandwidthClassUL-NR: a (0)
Item 1
BandParameters: nr (1)
nr
bandNR: 41
ca-BandwidthClassDL-NR: a (0)
featureSetCombination: 1
supportedBandwidthCombinationSet: c0 [bit length 2, 6 LSB pad bits, 11.. .... decimal value 3]
appliedFreqBandListFilter: 1 item
Item 0
FreqBandInformation: bandInformationNR (1)
bandInformationNR
bandNR: 41
supportedBandCombinationList-v1540: 2 items
Item 0
BandCombination-v1540
bandList-v1540: 1 item
Item 0
```


From here, It clearly shows it does not support n77

Thanks again for your help
Hendry

**Reply by herlesupreeth** [reply]
I dont see band 77 supported in the above capabilities

**Comment by luckybird77** [comment]
On a different subject, have you had experience with Google pixel 8 not able to attach with 999-70 but can with 00101 ? if you have insight feel free to let me know. thanks !

**Reply by herlesupreeth** [reply]
Unfortunately, I have no information related to Pixel devices. Try searching in the discussions/issues. I remember some other community member also facing issues with Pixel when using private PLMN

**Reply by gabri94** [reply]
Pixel will not work, check my comments above.

**Comment by luckybird77** [comment]
Thanks Herle

Yes thanks also for directing me to check the UE capabilities .. I did
confirm with Apple support that my particular model did not support n77

Hendry

**Comment by luckybird77** [comment]
Thanks Gabriele, I saw you posted this :

With a Magisk patch (need a rooted phone) I was able to override the DB and
enable PLMN 99901 on Pixel 6.
Unfortunately there doesn't seem to be a way without root.

do you have instructions or video showing how to root this and override the
DB

Hendry

**Reply by gabri94** [reply]
Not really, I haven't done it myself.
